packages/yew/src/virtual_dom/vlist.rs

```diff
@@ -175,52 +177,68 @@ mod feat_ssr {
-                    } else {
-                        let mut rest_child_furs: FuturesUnordered<_> = rest_child_furs.collect();
-                        async move { while rest_child_furs.next().await.is_some() {} }
-                            .right_future()
-                    };
-
-                    let transfer_fur = async move {
-                        // Render first child to parent buffer directly.
-                        first_child
-                            .render_into_stream(w, parent_scope, hydratable)
-                            .await;
-
-                        // Transfer results to parent writer.
-                        for mut r in child_streams {
-                            while let Some(next_chunk) = r.next().await {
-                                w.write(next_chunk.into());
+                _ => {
+                    async fn render_child_iter<'a, I>(
+                        mut children: I,
+                        w: &mut BufWriter,
+                        parent_scope: &AnyScope,
+                        hydratable: bool,
+                    ) where
+                        I: Iterator<Item = &'a VNode>,
+                    {
+                        let buf_capacity = w.capacity();
+                        let mut w = w;
+                        while let Some(m) = children.next() {
+                            let child_fur = async move {
+                                // Rust's Compiler does not release the mutable reference to
+                                // BufWriter until the end of the loop, regardless of whether an
+                                // await statement has dropped the child_fur.
+                                //
+                                // We capture and return the mutable reference to avoid this.
+
+                                m.render_into_stream(w, parent_scope, hydratable).await;
+                                w
+                            };
+                            pin_mut!(child_fur);
+
+                            match poll!(child_fur.as_mut()) {
+                                Poll::Pending => {
+                                    let (mut next_w, next_r) = io::buffer(buf_capacity);
+                                    // Move buf writer into an async block for it to be dropped at
+                                    // the end of the future.
+                                    let rest_render_fur = async move {
+                                        render_child_iter(
+                                            children,
+                                            &mut next_w,
+                                            parent_scope,
+                                            hydratable,
+                                        )
+                                        .await;
+                                    }
+                                    // boxing to avoid recursion
+                                    .boxed_local();
```

Comment:
Out of curiosity, what happens this isn't boxed? If the issue is stack overflow, is this something tail recursion can fix?

Reply:
With Rust’s async/await design, it’s not possible to recursively call async functions without boxing.

https://rust-lang.github.io/async-book/07_workarounds/04_recursion.html